src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -17,123 +21,170 @@ class MSMonitor(MSCore):
             requests = self.reqmgr2.getRequestByStatus([reqStatus], detail=False)
             self.logger.debug('+++ monit found %s requests in %s state',
                               len(requests), reqStatus)
 
-            # FIXME: completion threshold should come either from unified
-            # or campaign configuration, e.g.
-            # thr = self.unifiedConfig().get('completedThreshold', 100)
-            thr = 100
-
-            requestStatuses = {}  # keep track of request statuses
+            # get list of all transfer records
+            transferRecords = [d for d in self.getTransferInfo('ALL_DOCS')]
+            # keep track of completed workflows
+            completedWorkflows = []
+            # keep track of request and their new statuses
+            reqStatus = {}
+            # main logic
             for reqName in requests:
-                req = {'name': reqName, 'reqStatus': reqStatus}
-                self.logger.debug("+++ request %s", req)
-                # obtain status records from couchdb for given request
-                transferRecords = self.getTransferRecords(reqName)
-                if not transferRecords:
+                self.logger.debug("+++ request %s", reqName)
+                # for us workflow name is the same as request name
+                wname = reqName
+                # obtain transfer records for our request
+                transfers = []
+                for rec in transferRecords:
+                    if reqName == rec['workflowName']:
+                        transfers = rec['transfers']
+                        break
+                if not transfers:
                     continue
-                completion = self.completion(transferRecords)
                 # if all transfers are completed
                 # move the request status staging -> staged
-                if completion == thr:
+                if self.completion(transfers):
                     self.logger.debug(
-                        "+++ request %s all transfers are completed", req)
-                    self.change(req, 'staged', '+++ monit')
+                        "+++ request %s all transfers are completed", reqName)
+                    completedWorkflows.append(wname)
+                    reqStatus[reqName] = 'staged'
                 # if pileup transfers are completed AND
                 # some input blocks are completed,
                 # move the request status staging -> staged
-                elif self.pileupTransfersCompleted(transferRecords):
+                elif self.completion(transfers, pileup=True):
                     self.logger.debug(
-                        "+++ request %s pileup transfers are completed", req)
-                    self.change(req, 'staged', '+++ monit')
+                        "+++ request %s pileup transfers are completed", reqName)
+                    completedWorkflows.append(wname)
+                    reqStatus[reqName] = 'staged'
```

Comment:
It will always be to the same status, no need to keep it as a dict

Reply:
I'm not following here. First we loop over all requests we obtained from ReqMgr2. So, my understanding that each can progress independently from each other. Therefore each can carry its own status. One may be staged and another not. That's why I keep track between request names and their statuses.

Reply:
It could be. However, in the end you only have two options here:
a) request remains in the same status (doesn't matter what is that)
b) request moves to the next status, which is always `staged`